ratis-server/src/main/java/org/apache/ratis/server/impl/ServerState.java

```diff
@@ -182,17 +185,19 @@ void start() {
   }
 
   private static RaftLog initRaftLog(RaftGroupMemberId memberId, RaftServerImpl server, RaftStorage storage,
-      Consumer<LogEntryProto> logConsumer, long lastIndexInSnapshot, RaftProperties prop) throws IOException {
+      Consumer<LogEntryProto> logConsumer, LongSupplier snapshotIndexSupplier,
+      RaftProperties prop) throws IOException {
     final RaftLog log;
     if (RaftServerConfigKeys.Log.useMemory(prop)) {
-      log = new MemoryRaftLog(memberId, lastIndexInSnapshot, prop);
+      log = new MemoryRaftLog(memberId, snapshotIndexSupplier, prop);
     } else {
       log = new SegmentedRaftLog(memberId, server,
           server.getStateMachine(),
           server::notifyTruncatedLogEntry,
           server::submitUpdateCommitEvent,
-          storage, lastIndexInSnapshot, prop);
+          storage, snapshotIndexSupplier, prop);
     }
+    long lastIndexInSnapshot = snapshotIndexSupplier.getAsLong();
     log.open(lastIndexInSnapshot, logConsumer);
```

Comment:
snapshotIndexSupplier.getAsLong() is implemented by StateMachine so that it could be expensive.  Let's use log.getSnapshotIndex().
```
    log.open(log.getSnapshotIndex(), logConsumer);
```